File: backend/models/roles.py
"""Canonical role names and the grantable role vocabulary.

Roles are stored as rows in ``user_role`` (one row per grant) and are compared
as strings throughout the application. This module is the single source of
those strings, mirroring the relationship-type-constants convention in
``backend/graph/queries/relationships.py``: no role literal appears anywhere
else in application code.

Two roles are deliberately *not* constants of the storable kind:

* ``registered`` is implicit in holding an account — every authenticated caller
  has it, and it is never written to ``user_role``. It exists here only so the
  permission matrix in ``docs/architecture/roles-and-permissions.md`` § 2 can be
  named in code and docs.
* ``anonymous`` is the absence of authentication: an empty role set. There is no
  constant for it — code tests ``if user is None`` or an empty set, never a
  magic string.

See ``docs/architecture/roles-and-permissions.md`` § 1 and ADR-037.
"""

from __future__ import annotations

from typing import Final

#: Fragment tagging, annotation, and peer review (the Phase 1 role, unchanged).
EDITOR: Final[str] = "editor"

#: Blog authoring — deliberately distinct from :data:`EDITOR` (Component 16).
AUTHOR: Final[str] = "author"

#: Corpus management, user/role management, moderation.
ADMIN: Final[str] = "admin"

#: Roles that can be granted by an admin and stored in ``user_role``.
#: ``registered`` is implicit and ``anonymous`` is the absence of a session, so
#: neither appears here. A future ``translator`` role (ADR-006) is added here.
GRANTABLE_ROLES: Final[frozenset[str]] = frozenset({EDITOR, AUTHOR, ADMIN})

File: frontend/src/services/roles.ts
/**
 * Canonical role names — the frontend mirror of `backend/models/roles.py`.
 *
 * Roles are grants stored as `user_role` rows and compared as strings. This
 * module is the single source of those strings on this side of the wire, for
 * the same reason the backend module is on its side: no role literal should
 * appear anywhere else.
 *
 * `registered` is implicit in holding an account and is never stored, and
 * `anonymous` is the absence of a session — neither has a constant, here or on
 * the backend. See `docs/architecture/roles-and-permissions.md` § 1 / ADR-037.
 *
 * Nothing here is a permission boundary. Roles are enforced by the API on every
 * call; the frontend reads them only to decide what to *offer*.
 */

/** Fragment tagging, annotation, and peer review. */
export const EDITOR = 'editor';

/** Blog authoring — deliberately distinct from {@link EDITOR} (Component 16). */
export const AUTHOR = 'author';

/** Corpus management, user/role management, moderation. */
export const ADMIN = 'admin';

/** Roles an admin can grant, mirroring the backend's `GRANTABLE_ROLES`. */
export const GRANTABLE_ROLES = [EDITOR, AUTHOR, ADMIN] as const;

export type Role = (typeof GRANTABLE_ROLES)[number];

/**
 * The roles that reach the Editorial surfaces.
 *
 * Admin holds every editorial capability (the permission matrix, § 2), so it is
 * enumerated rather than implied: an admin-passes-everything rule would also
 * silently grant the author-only capabilities, which the matrix does not do.
 * This is the frontend counterpart of `require_role(EDITOR, ADMIN)`.
 */
export const EDITORIAL_ROLES: readonly string[] = [EDITOR, ADMIN];
